sqlglot/parser.py

```diff
@@ -632,32 +674,64 @@ def _parse_value(self):
         self._match_r_paren()
         return self.expression(exp.Tuple, expressions=expressions)
 
-    def _parse_cte(self):
+    def _parse_with(self):
         if not self._match(TokenType.WITH):
             return self._parse_select()
 
+        recursive = self._match(TokenType.RECURSIVE)
+
         expressions = []
 
         while True:
-            recursive = self._match(TokenType.RECURSIVE)
-            alias = self._parse_function() or self._parse_id_var()
+            expressions.append(self._parse_cte())
 
-            if not alias:
-                self.raise_error("Expected alias after WITH")
+            if not self._match(TokenType.COMMA):
+                break
 
-            if not self._match(TokenType.ALIAS):
-                self.raise_error("Expected AS after WITH")
+        this = self._parse_statement()
 
-            expressions.append(self._parse_table(alias=alias))
+        if "with" not in this.arg_types:
```

Comment:
These next few lines felt pretty nasty to me.
Let me know if you can think of a cleaner of doing this @tobymao